Replaced Token->validate with Validator class handling token validation

dannydorfel opened the pull request:
@lcobucci I kept the responsibility for validating in the claim, since they are like value objects and know best how to validate themselves. I'm not sure where to put the error message. In the claim itself? 

Also documentation still needs update in this PR. 

Your comments please?

KennedyTedesco commented:
Maybe the `validate()` method could be:

```
/**
 * @param Token $token
 * @param ValidationData $data
 * @return bool
 */
public function validate(Token $token, ValidationData $data): bool
{
    foreach ($this->getValidatableClaims($token) as $claim) {
        if ($data->has($claim->getName()) && ! $claim->validate($data)) {
            return false;
        }
    }

    return true;
}
```

KennedyTedesco commented:
Oops. Nevermind :smile: `getErrors()` explains everything.

dannydorfel commented:
Ah, I was already writing my explanation :-)

dannydorfel commented:
This PR is ment as a start to separate the validation concern from the token and add error messages. As mentioned in #72

KennedyTedesco commented:
Just another option:

```
/**
 * @param Token $token
 * @param ValidationData $data
 * @return bool
 */
public function validate(Token $token, ValidationData $data): bool
{
    $this->errors = [];

    foreach ($this->getValidatableClaims($token) as $claim) {
        if ($data->has($claim->getName()) && !$claim->validate($data)) {
            $this->errors[$claim->getName()] = false;
        }
    }

    return count($this->errors) === 0;
}
```

dannydorfel commented:
:+1: Ik like the improvement.

Op do 4 feb. 2016 19:53 schreef Kennedy Tedesco <notifications@github.com>:

> Just another option:
>
> /**
>  * @param Token $token
>  * @param ValidationData $data
>  * @return bool
>  */
> public function validate(Token $token, ValidationData $data): bool
> {
>     $this->errors = [];
>
>     foreach ($this->getValidatableClaims($token) as $claim) {
>         if ($data->has($claim->getName()) && !$claim->validate($data)) {
>             $this->errors[$claim->getName()] = false;
>         }
>     }
>
>     return count($this->errors) === 0;
> }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lcobucci/jwt/pull/73#issuecomment-179999141>.
>

lcobucci commented:
Awesome guys! I think we're going to a good way. 

After thinking a lot about those things I think that we can be a bit more radical :smile: : 

* get rid of the `Claim` objects (they are adding a bit of complexity that we may not need when removing this responsibility from the token);
* make the `Validator` a wrapper of an awesome validation library (like respect/Validation) 
* introduce a `ValidatorBuilder` to allow users to configure the values they want to use to create a validator (with a pre-configured set of constraints for the default claims) 

The benefits of that would be:

* simpler API (just use a validator instead of creating our own) 
* allowing the validation of private claims (on an easy way) #29

What you think?

lcobucci commented:
Maybe we could ping @henriquemoody or @alganet to get their input as well

Ocramius commented on src/Validator.php:
```diff
@@ -0,0 +1,63 @@
+            }
+        }
+
+        return count($this->errors) === 0;
```
Can be simplified to `! $this->errors`

dannydorfel replied:
I would disagree. I rather be explicit and not use the convenience (flaws) of the PHP language.

Ocramius replied:
At this point, this is a local variable, and you can safely replace an `O(n)` with an `O(1)` operation.

dannydorfel replied:
You are right and I know very well it's possible to negate the array, but I would like to advocate being explicit. Using ! on a variable would normally imply a boolean, but in PHP you can use it on a lot of different situations. For me, that is more or less like goto: just because we have goto in php doesn't mean we have to use it. Considering, using ```return empty()``` would be the best choice since we know it's an array.

Ocramius replied:
`empty()` is also fine, yes. `! $arr` is perfectly fine in PHP semantics (explicit too, to be honest)

dannydorfel replied:
:smile:

Ocramius commented on src/Validator.php:
```diff
@@ -0,0 +1,63 @@
+    /**
+     * @return array
+     */
+    public function getErrors(): array
```
This API is kinda bad: stateful error storage. You should instead simply return a `ValidationResult` that encapsulates the coupled state of `valid + errors`

dannydorfel replied:
I agree. This is work in progress and minimal working version, not a final version. Do you have a link showing me how we could improve?

Ocramius replied:
at the end of `validate()`, `return new TokenValidationResult($token, $validationData, $success, $errors)`

dannydorfel replied:
:+1: Will start on that, hopefully this weekend

henriquemoody replied:
I did not see @Ocramius' comment before my comment.

That would be nice as well :+1:

KennedyTedesco replied:
Great! :+1:

Ocramius commented on src/Validator.php:
```diff
@@ -0,0 +1,63 @@
+ */
+class Validator
+{
+    protected $errors = [];
```
`private`

dannydorfel replied:
:+1:

Ocramius commented on src/Validator.php:
```diff
@@ -0,0 +1,63 @@
+ *
+ * @author Danny Dörfel <danny.dorfel@gmail.com>
+ */
+class Validator
```
Missing an interface for it?

dannydorfel replied:
Would be a good addition

Ocramius commented on src/Validator.php:
```diff
@@ -0,0 +1,63 @@
+     * @param Token $token
+     * @return Generator
+     */
+    private function getValidatableClaims(Token $token)
```
Add ` : Generator` return hint

dannydorfel replied:
:+1:

Ocramius commented on src/Validator.php:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    /**
+     * Yields the validatable claims
```
Docblock can be entirely skipped

Ocramius commented:
About @lcobucci's suggestions:

 > * get rid of the `Claim` objects (they are adding a bit of complexity that we may not need when removing this responsibility from the token);

Depends on whether they are exposed or not

 > * make the `Validator` a wrapper of an awesome validation library (like respect/Validation)

Sounds like overkill/implementation detail. The current validation is sufficient and very simple, IMO. The only thing to be fixed is the stateful validator (really bad!)

 > * introduce a `ValidatorBuilder` to allow users to configure the values they want to use to create a validator (with a pre-configured set of constraints for the default claims) 

In that case, having just a validator interface and letting the user implement the validator seems more reasonable for a first version, or else you might be getting out of the 80/20 use-cases. Don't overdo it, IMO.

henriquemoody commented on src/Validator.php:
```diff
@@ -0,0 +1,63 @@
+     */
+    public function validate(Token $token, ValidationData $data): bool
+    {
+        $this->errors = [];
```
I thought about creating a JWT rule on Respect\Validation sometimes. Of course we can do that but we need to write too much code from this library on our library then the chances of breaking the API is higher than using a single point of failure, that's why this feature is so useful. Also I don't thing it's fair to force someone to install one more library to have a fature that's simple for to have here.

About this implementation, IMHO, I would say that it can be better and the problem is on this line. If is there something I would get away from is to change the state of the object on each validation, and that what is happening here.

First of all, this method breaks SRP and secondly there is no need for such thing, not here. This method should only return `true` or `false` and that's all.

So, how can we get the errors and still not change the state of the object?

You can do what we do on Validation: Have a method to `assert()` the claims, throwing an exception when one claim or more are invalid then reuse the same method on `validate()` with a try/catch statement and return `true` if not exception is thrown or `false` if an exception is thrown, so you don't break DRY.

lcobucci commented:
About @Ocramius and @henriquemoody comments:

> Depends on whether they are exposed or not

They're not exactly exposed. `Token#getHeaders()` and `Token#getClaims()` returns a list of them however it's way easier to just use `Token#getHeader($name)` and `Token#getClaim($name)` that hides the usage of `Claim`.

> Sounds like overkill/implementation detail. The current validation is sufficient and very simple, IMO. The only thing to be fixed is the stateful validator (really bad!)

> In that case, having just a validator interface and letting the user implement the validator seems more reasonable for a first version, or else you might be getting out of the 80/20 use-cases. Don't overdo it, IMO.

> I thought about creating a JWT rule on Respect\Validation sometimes. Of course we can do that but we need to write too much code from this library on our library then the chances of breaking the API is higher than using a single point of failure, that's why this feature is so useful. Also I don't thing it's fair to force someone to install one more library to have a fature that's simple for you to have here.

I think need to explain my suggestion better :smile:

The most important thing here (and I'm glad that we all agree on this HAHAHA) is to introduce an interface for the validator. In my mind it would be something like:

```php
namespace Lcobucci\JWT;

use Lcobucci\JWT\Validation\Result;

interface Validator
{
    public function validate(Token $token):Result;
}
```

Our default implementation would receive all the validation constraints (instances of `Respect\Validation\Validator`) and the `Validator#validate()` would just validate all the constraints (basically using the `Respect\Validation\Validator::allOf()`) and create a `Result` from that.

That means that we need an easy way to configure our validator's constraint list and that's where the builder enters. It would be the glue point with the chosen validation lib.

It's interface would be something like:

```php
namespace Lcobucci\JWT\Validation;

use Respect\Validation\Validator as Constraint;

class Builder
{
    public function assertId($id):Builder {}
    public function assertIssuer($issuer):Builder {}
    // the same for the other default claims
    public function assertThat($claim, Constraint $constraint):Builder {}
    public function getValidator():Validator {}
}
```

I'm just considering the idea of adding another dependency because we wouldn't have to deal here with the creation of error messages and also that we would allow users to validate the data in the token with some different validators.

Ocramius commented:
> deal here with the creation of error messages

Delegating this means less clear error messages (usually)

dannydorfel commented:
So we're clear on the interface (my time went 'poof' this weekend, so I'm in debt ;-))

How about keeping the assert in the claim, if you would consider a claim a value object. My preference is still to have the claim throw an exception. The validator can handle it and give a Result object back. Extracting it does feel a bit bloated.

As for the ```getValidator()``` in the builder: would not be my choice. The builder is a TokenBuilder and not a TokenValidatorBuilder, but I'm struggling with a better plan. Looks to me like the current builder class is already doing too much: building the headers, building the claims, building the token. So would it be an idea to extract the claims buildup in a separate class. That way we can reuse it in that object in the ```Validator::__construct(Claims $claims)```?

lcobucci commented:
> So we're clear on the interface (my time went 'poof' this weekend, so I'm in debt ;-))

Relax @dannydorfel, weekends are magical creatures that disappears sometimes =)

> How about keeping the assert in the claim, if you would consider a claim a value object. My preference is still to have the claim throw an exception. The validator can handle it and give a Result object back. Extracting it does feel a bit bloated.

I'm just suggesting to remove the claim objects because IMHO it's already a bit bloated heheh we're having to have a `ClaimFactory` inside of the `Builder` and inside of the `Parser` because of things related to the validation of each claim.

> As for the getValidator() in the builder: would not be my choice. The builder is a TokenBuilder and not a TokenValidatorBuilder, but I'm struggling with a better plan. Looks to me like the current builder class is already doing too much: building the headers, building the claims, building the token. So would it be an idea to extract the claims buildup in a separate class? That way we can reuse it in that object in the Validator::__construct(Claims $claims)?

If you look closely to my snippet you'll see that this `Builder` is inside other namespace and it's just related to validation :smiley:

dannydorfel commented:
A yes. But what if you can have a general ClaimBuilder for the claims. It's output is the same and can be used in the ```TokenBuilder``` and in the ```Validator```. Feels more SRP and decoupled than duplicate builders for token and validator. So extracting it from the builder makes it less bloated and still use the Claim as single value class that knows it's name and (if applicable) its validation.

And as I mentioned, I like the ValueObject-like setup. So removing the Claim object does not feel like an improvement. And I share @henriquemoody 's opinion on adding another library. You only have a few claim types, so throwing an Exception with correct message would suffice.

```php
namespace Lcobucci\JWT\Claim;

class Builder
{
    /** @var Collection */
    private $claims;
    private $factory;

    public function __construct(Factory $factory)
    { ... }

    public function setAudience(string $audience): Builder
    { ... }

    public function getClaims() : Collection
    {
        return $this->claims;
    }
        
}
```

```php
namespace Lcobucci\JWT;

use Lcobucci\JWT\Claim\Collection;

class Validator implements ValidatorInterface
{
    private $claims;

    public function __construct(Collection $claims)
    {
        $this->claims = $claims;
    }

    public function validate(Token $token): ValidateResult
    {
        $result = new ValidateResult();
        foreach ($this->claims as $claim) {
            .... try/catch, Claim with Validatable interface throws assert exception
        }

        return $result;
    }
}
```

```php
namespace Lcobucci\JWT;

use Lcobucci\JWT\Claim\Collection;

class Builder
{
    public function __construct(
        Collection $claims,
        Parsing\Encoder $encoder = null
    ) {
        ...
    }
}
```

dannydorfel commented:
@lcobucci @Ocramius @henriquemoody I changed the validator using interfaces and throwing exceptions. Please review ;-)

henriquemoody commented on src/Validation/ResultsInterface.php:
```diff
@@ -0,0 +1,30 @@
+/**
+ * Basic interface describing validation results
+ */
+interface ResultsInterface
```
I would use `ResultInterface`.

henriquemoody commented on src/Validation/ResultsInterface.php:
```diff
@@ -0,0 +1,30 @@
+     *
+     * @return bool
+     */
+    public function valid(): bool;
```
I would use `isValid()`.

henriquemoody commented on src/Validation/ResultsInterface.php:
```diff
@@ -0,0 +1,30 @@
+     *
+     * @return array
+     */
+    public function errors(): array;
```
I would use `getErrors()`.

henriquemoody commented on src/Validation/Results.php:
```diff
@@ -0,0 +1,54 @@
+     * @param string $name
+     * @param string $message
+     */
+    public function addError(string $name, string $message)
```
I think you should inject the exception here instead of defining these two arguments, you can also specialize the exception class.

henriquemoody commented on src/Validation/Results.php:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * @return bool
+     */
+    public function isExpired(): bool
```
I don't thins it's necessary.

henriquemoody commented on src/Claim/LesserOrEqualsTo.php:
```diff
@@ -23,12 +24,18 @@ class LesserOrEqualsTo extends Basic implements Claim, Validatable
+        $name = $this->getName();
+        if ($data->has($name) && ($this->getValue() > $data->get($name))) {
+            throw new InvalidClaimException(
+                sprintf(
```
You're repeating yourself with `sprintf($message, $name, $data->get($name), $this->getValue())`

dannydorfel commented:
So if have changed the exceptions for the claims after @henriquemoody 's review.
I specialised the exception class, replaced the name and message parameter in the Results class. And I also added the RangeException, extending the InvalidClaimException, defining a more specific exception for less/equal or greater/equal mismatch.

I hope I'm not asking too much, but am I on the right track here?

dannydorfel commented:
Just a reminder to keep things alive @lcobucci . No worries, I think this PR is not to be rushed. I'm actually expecting more refactor / review rounds.

lcobucci commented:
Thanks @dannydorfel for your patience! Since yesterday evening I finally have internet connection at my new home so things will be faster now :smile: 

About this PR, you were able to solve the SRP violation however it's way more complex than what I thought it would be. I was also hoping to solve the #29 with this extraction but I don't see this happening in the path we're going.

For me it doesn't matter if we're going to use a 3rd party lib or not, my idea is just to make everything simple and extensible. I just don't want you to think that I'm not considering the work you've done, I just think that the path I have in mind is easier to implement, removes some unnecessary complexity that we currently have and also solves the SRP problem - but I'm just unable to convert everything in words now =/.

I'm not sure if I should translate what I'm thinking into another PR and we could compare both paths and pick the best one or try to explain it here.

What's your suggestion guys?

henriquemoody commented:
I will be glad to review another pull request and discuss what's the best path to follow, @lcobucci. In spite of all the work that @dannydorfel has been done since then, I think it will not be a problem to have more options.

dannydorfel commented:
I'm actually glad to read that your looking for a simpler solution @lcobucci. Since Ocramius started with "having just a validator interface and letting the user implement the validator". I actually started my commits with only moving to the validator, removing the SRP issue. So it feels like a huge change and a lot the '80/20 use-cases' Ocramius talked about.

I wouldn't mind discussing a minimal code needed to remove the SRP and keeping the possibilities for #29 open? How about starting with only the ValidatorInterface and leaving the claims to returning false?

lcobucci commented:
Closing this in favor of #75.
